4.0.0rc1

fhdk opened the pull request:
started working in incorporating the new status/mirrors files

Huluti commented:
With these new JSON files, the goal is to remove the **data** folder. So, we need to remove and replace all parts that use the files in the **data** folder.

fhdk commented:
@Huluti 
I have committed two small fixes to the master - hope you have seen them.

It is fixes for breaks when a country is not spelled correct and if the Custom mirror file does not exist.

I think they should be pushed to build don't you?

fhdk commented:
For this branch:

What is the plan with mirrors.json?
Are we saving it to `/etc/pacman.d/mirrors/mirrors.json` for reuse?
What conditions should initiate the mirrors.json be downloaded?
An argument to pacman-mirrors to initiate a redownload of mirrors.json?
On intervals like in pamac?
An elapsed time where last time is saved in pacman-mirrors.conf?

Kirek commented:
The data folder needs to stay, meaning that we need to deliver a list of mirrors within the package, we cannot completely rely that repo.manjaro.org is accessible. We don't have dozen of mirrors and then have a single point of failure with repo.manjaro.org
So the file at https://raw.githubusercontent.com/manjaro/manjaro-web-repo/master/mirrors.json (or whatever its it final location) needs to be copied somewhere to the package/hard disk during build time ( /etc/pacman.d/mirrors/mirrors.conf for example), and pacman-mirrors fallback to this file if the connection to repo.manjaro.org fails.

fhdk commented:
No SPOF agreed

Should it be saved in the current Custom mirrorfile format? without the response and last updated - which I also think should be removed from the Custom mirrorfile - but that is another thing.

Kirek commented:
I think mirrors.json can be downloaded every time pacman-mirrors is run, by default.
It can be saved with the same format as mirrors.json, so we can keep it simple: can't donwload the file from repo.manjaro.org? read the one locally, that has the same format.

fhdk commented:
Same format - I like that - how about the end user which might be looking after country specific files?

Are we supporting this?

Kirek commented:
No need to keep country specific files, the current mirrors.json file is perfectly human readable, and grouped by country.

fhdk commented:
I agree - just want to know :)

So the flow will be something like this

* mirrors.json is packaged so it exist in `/etc/pacman.d/mirrors` as mirrors.json
* pacman-mirrors fetch the file by http at every run
* pacman-mirrors overwrites existing mirrors.json
* if not possible to reach repo.manjaro.org fallback to local file

I think we could do the same for the `status.json` file fetch on every run.

What are we going to do with the data?

* Do we still need to rank the mirrors locally? Eg. a mirror in geolocated country could more responsive seen from that country than from the server.
* Are we going to do some sorting on the users selected branch state?
* Are we going to do a combined sort on state and response?

Kirek commented:
Yes, I think this will be a reasonable flow.

> Do we still need to rank the mirrors locally? Eg. a mirror in geolocated country could more responsive seen from that country than from the server.
Yes, we still need to ping the servers, we don't need to parse the state file as we already have state.json

> Are we going to do some sorting on the users selected branch state?
I'm not sure what you mean

> Are we going to do a combined sort on state and response?

If you mean state as last_sync, I think the current aproach is good, if the last_sync is reasonably good (les than 24h), we sort by response time.

fhdk commented:
I think I got the answers I need

MIRRORS_FILE = "/etc/pacman.d/mirrors/mirrors.json"
STATES_FILE = "/etc/pacman.d/mirrors/states.json"

* mirrors.json is packaged so it exist in /etc/pacman.d/mirrors as mirrors.json
* pm fetch mirrors.json by http at every run
* pm overwrites existing mirrors.json
* pm fetch status.json by http at every run
* pm overwrites existing status.json
* if not possible to reach repo.manjaro.org fallback to local file

* mirrors is ranked by pinging them instead of checking state file
* data presentation and sorting remains unchanged

Please evaluate if other approach is needed.

Huluti commented:
For me, the purpose of this change is also to be able to revert to a **classic version number**. We'll could normally rename the **pacman-mirrorlist** package to **pacman-mirrors** since it will no longer depend on the mirrors in the data folder.

fhdk commented:
Are we in agreement that to maintain the complete transition to json, we should be converted the Custom mirror file and save as `/var/lib/pacman-mirrors/custom.json`?

Kirek commented:
If we change Custom to custom.json we need to automatically migrate the contents.

> For me, the purpose of this change is also to be able to revert to a classic version number. We'll could normally rename the pacman-mirrorlist package to pacman-mirrors since it will no longer depend on the mirrors in the data folder.

We can use this update to split the package: 

pacman-mirrolist will only have the fallback mirrors.conf, at /usr/share/pacman-mirrors/mirrors.conf this will never be modified by the script, and will work as fallback if can't connect to repo.manjaro.org, this could be needed at live media or first boot, for example.

pacman-mirror will have the current script, and store the data to /var/lib/pacman-mirrors/ (that means the custom file, mirrors.json and status.json will be kept there). This way we remove all the data from /etc

Another thing about development, if you find more bugs for the current 3.2.x branch, push them there directly, as master will become somewhat unstable shortly :)

fhdk commented:
I am currently looking at a way to convert the Custom mirror file to json.

Of course I strive to hit the same format as mirrors.json but I can't.

I have been reading on python dictionary and wringling my brain to understand mirrors.json, to me it looks like a big dictionary with a bunch of unnamed country dictionaries with an unnamed mirrors dictionary with a named protocols dictionary with list data.

How did you create a json object with a bunch of country objects with a couple of mirror objects with a named protocols object with a list.

I have gotten to creating the country dictionary but how do I get it into a JSON object and not a list.

fhdk commented:
I have set constants in configuration.py to point to filenames while in dev.

Also in configuration is the final paths and as I read previous posts we are moving house-keeping to /var/lib/pacman-mirrors.

Am I correct in assuming this?

Kirek commented:
the mirrors.json format is actually quite weird, its like a dictionary inside a dictionary inside a dictionary and then a list of protocols?
Would be much simpler if it was just a list of dictionaries
[{
"country": "foo",
"url": "http://...",
"protocols": ["http"]
},
"country": "bar",
"url": "http://...",
"protocols": ["https"]
}]

In fact I think how the status.json is organized, but with extra data.

fhdk commented:
Agree - status.json is not quite there - the branches just have statuses 1,1,0 - would be nice if it followed the same conventions

I think this format for mirrors.json and status.json would be the optimal - I think if user selects specific countries - when mirrors.json is requested last_sync and branches are left empty and status.json that goes for it self
```
[{
    "foo-country": [
        {
            "mirror": "http://.......",
            "procotols": ["http"],
            "last_sync": "hh:mm",
            "branches": ["1","1","0"]
        },
        {
            "mirror": "http://.......",
            "procotols": ["http"],
            "last_sync": "hh:mm",
            "branches": ["1","0","0"]
        }
    ],
    "bar-country": [
        {
            "mirror": "http://.......",
            "procotols": ["http"],
            "last_sync": "01:09",
            "branches": ["0","0","0"]
        },
        {
            "mirror": "http://.......",
            "procotols": ["http"],
            "last_sync": "01:09",
            "branches": ["0","0","0"]
        }
    ]
}]

Huluti commented:
It's not good idea to have just one file: manjaro-web-repo also uses the file **mirrors.json**. Each hour, the server **pull** the repo to check updates. If the file **mirrors.json** has an update (a new mirror for example), this will create a conflict between last_sync and branches values.

Huluti commented:
> the mirrors.json format is actually quite weird, its like a dictionary inside a dictionary inside a dictionary and then a list of protocols?
> Would be much simpler if it was just a list of dictionaries
> 
> [{
> "country": "foo",
> "url": "http://...",
> "protocols": ["http"]
> },
> "country": "bar",
> "url": "http://...",
> "protocols": ["https"]
> }]
> In fact I think how the status.json is organized, but with extra data.

But I'm ok with the suggestion of Kirek. I'll do that.

fhdk commented:
> It's not good idea to have just one file: manjaro-web-repo also uses the file mirrors.json. Each hour, the server pull the repo to check updates. If the file mirrors.json has an update (a new mirror for example), this will create a conflict between last_sync and branches values.

I see - could it at least be possible to have it country wise as I suggested?
```
[{
    "foo-country": [
        {
            "mirror": "proto://.......",
            "procotols": ["http", "https", "ftp", "rsync"],
            "last_sync": "99:99",
            "branches": [
                {
                    "stable": "0",
                    "testing": "0",
                    "unstable": "0"
                }
            ]
        }
    ],
    "bar-country": [
        {
            "mirror": "proto://.......",
            "procotols": ["http", "https", "ftp", "rsync"],
            "last_sync": "99:99",
            "branches": [
                {
                    "stable": "0",
                    "testing": "0",
                    "unstable": "0"
                }
            ]
        }
    ]
}]
```

Huluti commented:
Not like Kirek's suggestion to match status.json?

> [{
> "country": "foo",
> "url": "http://...",
> "protocols": ["http"]
> },
> "country": "bar",
> "url": "http://...",
> "protocols": ["https"]
> }]

fhdk commented:
@Huluti @Kirek 
If you plan on splitting or renaming the package wouldn't it be appropriate to name it manjaro-mirrors - to avoid thinking it works with arch too?

fhdk commented:
@Huluti 
> It would facililtate an easy extraction of countries when a user is passing in -c country,country,country

If there must be two files then so it is
The optimal is the structure like the status.json but created country-wise - like mirrors.json is that possible?

```
[{
    "foo-country": [
        {"mirror-a": "://", "procotols": [], "last_sync": "", "branches": []},
        {"mirror-b": "://", "procotols": [], "last_sync": "", "branches": []}        
    ],
    "bar-country": [
        {"mirror-c": "://", "procotols": [], "last_sync": "", "branches": []},
        {"mirror-d": "://", "procotols": [], "last_sync": "", "branches": []}
    ]
}]
```

Kirek commented:
It's very easy to filter a list of dictionaries, for example in the structure I suggested it will be something like that: 
(mirror_list variable is a list of dictionaries in the current format of status.json)

Match a single country:
```
country_list = ['United_States']
result = [x for x in mirror_list
            if x['country'] in country_list]
```

Or various:
```
country_list = ['United_States','Germany']
result = [x for x in mirror_list if x['country'] in country_list]
```

or you can sort them easily too
```
from operator import itemgetter
newlist = sorted(mirror_list, key=itemgetter('country'))
```

fhdk commented:
@Kirek 
>It's very easy to filter a list of dictionaries, for example in the structure I suggested it will be something like that:
(mirror_list variable is a list of dictionaries in the current format of status.json)

Thanks for the guidance - much appreciated :+1:

fhdk commented:
Is this structure for data files acceptable or do you want something changed

https://github.com/fhdk/pacman-mirrors/blob/3.3-unstable-json/pacman_mirrors/configuration.py

philmmanjaro commented:
Can somebody explain to me what the **/dev** folder structure is for?

philmmanjaro commented:
877ba15 should be reverted. Next time please ask before you change the application name completely ...

fhdk commented:
I have not changed any name just asked as can be seen in previous posts. It was also triggered by some comments from @Huluti.

The request to review the configuration was also intended as a question on the matter raised earlier by @Huluti in regard to his intentions on package splitting renaming.

⁣Sendt af TypeApp ​

Den 13. feb. 2017 21.54, fra 21.54, "Philip Müller" <notifications@github.com> skrev:
>877ba15 should be reverted. Next time please ask before you change the
>application name completely ...
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/manjaro/pacman-mirrors/pull/61#issuecomment-279518882

fhdk commented:
It is for testing files read/write during dev. It is to be removed once the primary phase is over. It is a helper structure for me.

⁣Sendt af TypeApp ​

Den 13. feb. 2017 21.51, fra 21.51, "Philip Müller" <notifications@github.com> skrev:
>Can somebody explain to me what the **/dev** folder structure is for?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/manjaro/pacman-mirrors/pull/61#issuecomment-279518193

fhdk commented:
@philmmanjaro @Kirek @Huluti 
>We can use this update to split the package:
>
>pacman-mirrolist will only have the fallback mirrors.conf, at /usr/share/pacman-mirrors/mirrors.conf this >will never be modified by the script, and will work as fallback if can't connect to repo.manjaro.org, this >could be needed at live media or first boot, for example.
>
>pacman-mirrors will have the current script, and store the data to /var/lib/pacman-mirrors/ (that means >the >custom file, mirrors.json and status.json will be kept there). This way we remove all the data from >/etc

As I read the above this is the file/folder layout (snip from configuration.py)
```
    # etc
    CONFIG_FILE = "/etc/pacman-mirrors.conf"
    MIRROR_LIST = "/etc/pacman.d/mirrorlist"
    # pacman-mirrors
    MIRROR_DIR = "/var/lib/pacman-mirrors/"
    CUSTOM_FILE = MIRROR_DIR + "custom-mirrors.json"
    MIRROR_FILE = MIRROR_DIR + "mirrors.json"
    STATUS_FILE = MIRROR_DIR + "status.json"
    # special cases
    FALLBACK = "/usr/share/pacman-mirrors/mirrors.json"
    O_CUST_FILE = MIRROR_DIR + "Custom"
```
Is it OK I create a mock structure in the project root to support tests?
```
mock
    etc
        pacman-mirrors.conf
    usr
        mirrors.json
    var
        Custom
        mirrors.json
        custom-mirrors.json
        status.json
```

fhdk commented:
Found a way to parse existing mirrors.json
```
            for k in mirrors.keys():
                print("got key", k, "which maps to value", mirrors[k])
                for x in mirrors[k]:
                    print("got key", x, "which maps to value", mirrors[k][x])
                    for y in mirrors[k][x]:
                        print("got key", y, "which maps to value", mirrors[k][x][y])

```

Huluti commented:
Don't care about that, I have to change this directly in manjaro-web-repo.

fhdk commented:
On Tue, 2017-02-14 at 08:31 -0800, Hugo Posnic wrote:
> Don't care about that, I have to change this directly in manjaro-web-
> repo.

I know I raised the issue since I didn't understand how you did it, but
if it is too much trouble don't change it.

Huluti commented:
It's no much work, don't worry ;) I'll do that by tomorow.

fhdk commented:
@Huluti 
When you're at it - I have tiny request

In mirrors.json write the url without protocol
```
    "Australia": {
        "mirror.ventraip.net.au/Manjaro/": {
            "protocols": ["http"]
        },
        "manjaro.uberglobalmirror.com/": {
            "protocols": ["http"]
        },
        "manjaro.mirror.serversaustralia.com.au/": {
            "protocols": ["http"]
        }
```

fhdk commented:
>Since protocols is part of the structure in mirrors.json it would be nice if you wrote the url without protocol

On second thought it doesn't matter

The idea was since protocols is listed to test each protocol - but it is not necessarry to test all available protocols for response time. 

I will only test http and then write all protocols to pacmans mirrorlist using replace on 'http'

Huluti commented:
I've done the change: https://github.com/manjaro/manjaro-web-repo/commit/f2cf00dc451f3a80563dde00948626dd92ea672a.

fhdk commented:
@Huluti 
I have have a small issue

The last_sync value is always a string - except when its not - not a big deal but thought you should know
```
    {
        "branches": [
            -1,
            -1,
            -1
        ],
        "country": "Brasil",
        "last_sync": -1,
        "protocols": [
            "http"
        ],
        "url": "http://linorg.usp.br/manjaro/"
    },
    {
        "branches": [
            1,
            0,
            0
        ],
        "country": "Bulgaria",
        "last_sync": "11:48",
        "protocols": [
            "http",
            "https"
        ],
        "url": "https://mirrors.netix.net/manjaro/"
    }
```

fhdk commented:
I think I would like you to review the code - see if I overlooked something obvious - the usual dummie stuff

fhdk commented:
>        "country": "Brasil",
>        "last_sync": -1,
>        "protocols": [
>            "http"
>        ],
>        "url": "http://linorg.usp.br/manjaro/"

When "last_sync" is -1 instead of eg. "9999:99" my code breaks several places.
In the mirror_to_server_list - if last_sync is int -1 it set it to "9999:99"

fhdk commented:
I am ready for you guys to do a thorough review and test for whatever I missed

Huluti commented:
Thank you for you work! I'll take time tonight to make an entire review. ;)

Huluti commented on pacman_mirrors/httpfn.py:
```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python3
```
Replace by #!/usr/bin/env python. "python" is python3 by default.

fhdk replied:
check

Huluti commented on locale/pacman_mirrors.pot:
```diff
@@ -7,9 +7,9 @@
 msgid ""
 msgstr ""
-"Project-Id-Version: pacman-mirrors 3.1.0\n"
+"Project-Id-Version: pacman-mirrors 3.3b0\n"
```
Adjust version.

fhdk replied:
check - set at 3.3.0 - in the end you decide if it should be 4

Huluti commented on pacman_mirrors/graphical_ui.py:
```diff
@@ -36,8 +35,11 @@
-    def __init__(self, server_list):
-        Gtk.Window.__init__(self, title=txt.I_TITLE)
+    def __init__(self, server_list, random):
+        if not random:
```
Could you inverse conditions like this? It's more natural.

fhdk replied:
check - you're right

Huluti commented on pacman_mirrors/mirror.py:
```diff
@@ -0,0 +1,84 @@
+            branches = [0, 0, 0]
+        if country not in self.countrylist:
+            self.countrylist.append(country)
+        # workaround for negative integer in status.json
```
It's not a workaround. For me, it's fine: just the translation of a parameter to get what we want.

fhdk replied:
check - line is now
# translate negative integer in status.json

Huluti commented on pacman_mirrors/mirrorfn.py:
```diff
@@ -0,0 +1,69 @@
+        """
+        workitem = mirror
+        handle.write("## Country       : {}\n".format(workitem["country"]))
+        # TODO: approval to remove useless lines
```
@Kirek You are better able to judge than me

Kirek replied:
Yes, feel free to remove the commented lines

Huluti commented on pacman_mirrors/pacman_mirrors.py:
```diff
@@ -128,7 +116,7 @@ def command_line_parse(self):
                             metavar=txt.SECONDS,
                             help=txt.HLP_ARG_TIMEOUT)
-        parser.add_argument("--no-update",
+        parser.add_argument("--no_update",
```
Not a good idea to change parameters.

fhdk replied:
You're absolute right - had no intentions regarding that - thanks for spotting it :)
But I think I spotted it before you because my version is correct.

fhdk replied:
You're right - don't know how - check

Huluti commented on pacman_mirrors/pacman_mirrors.py:
```diff
@@ -165,64 +154,131 @@ def command_line_parse(self):
+            else:
+                print(".: {} {}".format(txt.WRN_CLR, txt.INF_NO_SELECTION))
+                print(".: {} {}".format(txt.WRN_CLR, txt.INF_NO_CHANGES))
+        else:
```
These two lines (253-254) aren't useful.

fhdk replied:
I humbly disagree.
The only way a user can exit from TUI is by clicking OK - so in that case it is nice to know nothing has changed.

Huluti commented on pacman_mirrors/txt.py:
```diff
@@ -58,38 +58,51 @@
+INF_CONVERT_MIRROR_FILE = _("Converting custom mirror file to new format")
 INF_CUSTOM_MIRROR_FILE = _("Custom mirrors file")
 INF_DOES_NOT_EXIST = _("doesn't exist.")
+INF_DOWNLOAD_MIRROR_FILE = _("Downloading mirrors from manjaro.org")
```
Replace manjaro.org by repo.manjaro.org.

fhdk replied:
of course - check

fhdk commented:
Question regarding python versioning in py files.

If a user - for some unknown reason - has changed their default python interpreter to python2 pacman-mirrors will fail miserably.

```
$ sudo ln /usr/bin/python2 /usr/bin/python --force
$ sudo pacman-mirrors -g
Traceback (most recent call last):
  File "/usr/bin/pacman-mirrors", line 23, in <module>
    from pacman_mirrors import pacman_mirrors
ImportError: No module named pacman_mirrors
```
Wouldn't the overall definition in the py files prevent such failure?

philmmanjaro commented:
I assume #62 will make this obsolete. However, @Huluti, @Kirek, was any of this code tested yet? If needed we may create a **pacman-mirrors-dev** package to have a wider audience ...

fhdk commented:
It could - I just didn't want to mix it all in in the same PR :)

Kirek commented on pacman_mirrors/pacman_mirrors.py:
```diff
@@ -165,64 +156,129 @@ def command_line_parse(self):
-            self.max_wait_time = args.timeout
+        if args.country and not args.geoip:
+            self.config["only_country"] = args.country.split(",")
 
```
You're right, they are mutually exclusive, but I think a better approach will be to exit and inform the user they can't use --country and --geoip at the same time.

fhdk replied:
Before I joined the project the output of pacman-mirrors -h was this.
It states that -c is ignored if --geoip is supplied. I just moved the internal check so it occurs earlier
```
usage: pacman-mirrors [-h] [-g] [-m {rank,random}]
                      [-b {stable,testing,unstable}] [-c COUNTRY] [--geoip]
                      [-d PATH] [-o FILE] [-t SECONDS] [--no-update] [-i] [-v]
                      [-q] [-f DIGIT]

optional arguments:
  -h, --help            show this help message and exit
  -g, --generate        Generate mirrorlist
  -m, --method {rank,random}
                        Generation method
  -b, --branch {stable,testing,unstable}
                        Branch name
  -c, --country COUNTRY
                        Comma separated list of countries, from which mirrors
                        will be used
  --geoip               Get current country using geolocation. Ignored if
                        '-c/--country' is supplied
  -d, --mirror_dir PATH
                        Mirrors list path
  -o, --output FILE     Output file
  -t, --timeout SECONDS
                        Maximum waiting time for server response
  --no-update           Don't generate mirrorlist if 'NoUpdate = True' in the
                        configuration file
  -i, --interactive     Generate custom mirrorlist
  -v, --version         Print the pacman-mirrors version
  -q, --quiet           Quiet mode - less verbose output
```

Kirek commented on tests/test_pacman_mirrors.py:
```diff
@@ -16,137 +16,81 @@
-                                  "-c", "Germany",
-                                  "-m", "random",
-                                  "-d", "data/mirrors"]):
+                                  "-c", "Germany"]):
```
not using -m random with this tests may cause them to take a lot of time.

fhdk replied:
I am working on splitting internals down to manageable testable units..
The actual server_list generation runs in a separate function which is omitted in everything else but the test run of the entire app.

Test entire app
```
with unittest.mock.patch("sys.argv",
                         ["pacman-mirrors",
                          "--quiet",
                          "-m", "random"]):
    app = pacman_mirrors.PacmanMirrors()
    app.config = app.load_conf()
    app.command_line_parse()
    app.load_mirror_file()
    app.validate_custom_config()
    app.validate_country_selection()    
    app.gen_server_lists()
        if app.interactive:
            app.gen_mirror_list_interactive()
        else:
            app.gen_mirror_list_common()
```
other don't include the lines ranking/randomizing and mirrorlist creation
```
app.gen_server_lists()
if app.interactive:
    app.gen_mirror_list_interactive()
else:
    app.gen_mirror_list_common()
```

Kirek commented:
When using method "random" it's ignoring -c and may be ignoring --geoip too 
It doesn't ignore it when using interactive.

Kirek commented:
Now when using a country that doesn't exist, instead of informing the user of the error, list the available countries and quit, it continues using all servers. This change should be reverted.

fhdk commented:
> Now when using a country that doesn't exist, instead of informing the user of the error, list the available countries and quit, it continues using all servers. This change should be reverted.

check
```
for country in selection:
    if country not in countrylist:
        print(".: {} {}{}: {}: '{}'.\n\n".format(txt.WRN_CLR,
            txt.INF_OPTION,
            txt.OPT_COUNTRY,
            txt.INF_UNKNOWN_COUNTRY,
            country))
            exit(1)  # exit gracefully if validation fail
return True
```

fhdk commented:
This is driving me nuts
```
    if self.config["method"] == "rank":
TypeError: 'bool' object is not subscriptable
```
What am I missing?

I get that the result of the operation is a bool value but why does it break?

fhdk commented:
@philmmanjaro @Kirek @Huluti 
While I are putting a new test suite together - the suggestion of a pacman-mirrors-dev package sounds good to me.

While I have put a lot of effort in getting this right - there still might be something I missed.

Huluti commented:
I like this idea also. This will avoid the recent problems. Good work @fhdk!

fhdk commented:
@Kirek @Huluti @philmmanjaro 
When is it possible to make dev package?

philmmanjaro commented:
I've added **pacman-mirrorlist-dev** to our [core repo](https://github.com/manjaro/packages-core/commit/b05c279c0e85a4fdfbad77ee4caf3c236560fdcc). How far are we now with this?

fhdk commented:
I have been through the code today - fixing minor details and I think it's good.

philmmanjaro commented:
Ok, keep me posted and I'll update the package.

fhdk commented:
Should I change the code from "dev"? to ""

It was my workaround to make tests work and it has no place in build

fhdk commented:
Just changed "dev" to ""
Most travis tests will fail as it cannot find the right pacman-mirrors.conf

fhdk commented:
@philmmanjaro 
Thanks :smirk_cat:  - The build unfortunately executed before I could change the variable.
Does it rebuild when I change a file? Otherwise I would appreciate if you could bump the package.

philmmanjaro commented:
@fhdk: bumping a unreleased package makes no sense at all. You can use **3.2.90** or some like that to indicate a development release. The package is now at 6df3268.

fhdk commented:
sorry if I sound ignorant - I have no experience with this kind of development - so please bear with me :smile:

philmmanjaro commented:
@fhdk: all good. Keep on going. However take a look at [sematic versioning](http://semver.org/). @Kirek; @Huluti please also review the current version and ping me when **v3.3.0** is ready to be released.

fhdk commented:
1. I have looked at the semantic versioning. As this version completely alters the way files are handled, abandons the current `mirrors` directory and convert, if present, the users Custom mirrorfile - am I correct in understanding this as breaking backwards compatibility? The current arguments for `mirror_dir` and `output` - I have not dived into their functionality but it seems to me they are becoming obsolete. What do You think of it? I will suggest a release v4.0.0 and a development version v3.9.9. 

2. How do I handle this, when trying to install the dev package?
```
resolving dependencies...
looking for conflicting packages...
:: pacman-mirrorlist-dev and pacman-mirrorlist are in conflict. Remove pacman-mirrorlist? [y/N] y
error: failed to prepare transaction (could not satisfy dependencies)
:: pacman: removing pacman-mirrorlist breaks dependency 'pacman-mirrorlist>=20130418'
```

philmmanjaro commented:
@fhdk: well, in regard of development you can do this:

* either use words like **3.3.0-alpha1**, **3.3.0-alpha2**, **3.3.0-beta1**, **3.3.0-rc1**, **3.3.0**
* or you can use versions like **3.2.70**, **3.2.71**, **3.2.80**, **3.2.90**, **3.3.0**.

If you have changed something dramatically the version has to be bumped. That we have to verify. Revison2 should work now. I've updated the **provides** value.

Huluti commented:
I'll do a rereading this afternoon.

Huluti commented:
I like how you have split the files. The code is easier to read. I have a suggestion that could get the code a bit clearer. You could use simple functions instead of creating class with just static methods. You can take a look at this link http://stackoverflow.com/a/18262662. It's a common practice in Python. However, very great work. 👍 I'am still reading the code.
@Kirek what do we do with the current content of the **data** folder?

fhdk commented:
Just read the referenced link - always ready to learn something new :smile: 

I am very familiar with namespaces :smile: and kind of fond of them too.

Is it the filename which sort-of sets the namespace?

Huluti commented:
I will want to say yes but I'm not sure. For me it's the opposite, I'm not very familiar with namespace. 😄

fhdk commented:
The post you referenced contained a link to http://2ndscale.com/rtomayko/2004/the-static-method-thing

I have marked the bold place - thats what made me see the logic in your previous proposal.
Quote:
> Here’s the idiomatic translation to Python of the Java example above. The following is in a module file named X.py:
> 
> message = "Hello"
> def say_hello():
>     print message
> And calling the function from another module (in Main.py):
> 
> import X
> X.say_hello()
> Now let’s look at an example of what not to do in Python:
> 
> class X:
>     message = "Hello"
>     def say_hello(cls):
>         print X.message
>     say_hello = classmethod(say_hello)
> Calling the above from another module looks like this:
> 
> import X
> X.X.say_hello()
> Here’s why: in Python, each module is itself an object without having to define an explicit class. **A single module in Python maps to a single .py file on disk**. Modules may contain functions, attributes (variables), multiple class definitions, etc. There is no equivalent of the Python module in Java - each .java file must have a single top-level class that matches the filename. Imports in Java work at the class level; imports in Python work at the module level.
>

Huluti commented:
I see, thank's. 👍 So have you an idea of which static methods should be just functions or not?

fhdk commented:
I have a good idea and I am looking at it right now

fhdk commented:
@philmmanjaro 
Sorry - corrected

fhdk commented:
@Huluti 
I have just pushed the changes so if you would be so kind to have a look?

Huluti commented:
The PEP-8 recommends absolute imports:

> Absolute imports are recommended, as they are usually more readable and tend to be better behaved (or at least give better error messages) if the import system is incorrectly configured (such as when a directory inside a package ends up on sys.path ).
https://www.python.org/dev/peps/pep-0008/

Can you make a try?

Huluti commented:
1e9a75abd91c3f4e83ce516048a77b417dbe6346: great commit!

fhdk commented:
@Huluti 
had a look at the link in your comment - I pushed some pep fixes
also tried with import filename but it gave me errors

so I am sticking to `from . import filename` which is also approved in linked document :smile:

philmmanjaro commented:
All right then. Package is now at 8d45cee

Huluti commented:
@philmmanjaro we have to think about renaming the package and use the right version number (instead of date).

fhdk commented:
When you decide to release this package into the wild :) please consider adding the current mirrors.json to the package.

If everything fails the app expects a fallback file at `/usr/lib/pacman-mirrors/mirrors.json`. I don't know if it should be this location but it is easily changed in `configuration.py`.

The app does not use any files from the data folder so the folder can actually be removed from the dist-package.

philmmanjaro commented:
@Huluti: well, some packages still need the package **pacman-mirrorlist**. If we release **pacman-mirrors** some other packages need to been adopted. Also a **mirrorlist** needs to be generated so **pacman** is functional. Lets finish this version and then we check on how we make it possible.

philmmanjaro commented:
@fhdk: If that **data** folder is not needed, then you might need to adjust **setup.py** accordingly:

https://github.com/manjaro/pacman-mirrors/blob/master/setup.py#L42
https://github.com/manjaro/pacman-mirrors/blob/master/setup.py#L55

philmmanjaro commented:
Package is now at 6362571

fhdk commented:
Looking at setup.py.
What is the best approach to have the, at any given time, a current mirrors.json to include in the package?

Would a function which fetches it from repo.manjaro.org be ok?

philmmanjaro commented:
@fhdk: since we have it in github, it would be good to get it from there.

```
  https://github.com/manjaro/manjaro-web-repo/raw/master/mirrors.json
  install -Dm644 "${srcdir}/mirrors.json" "${pkgdir}/usr/share/pacman-mirrors/mirrors.json"
```

fhdk commented:
@philmmanjaro 
I have updated the setup.py 

- do not package the country files
- fetch the current mirrors.json and save it to data/mirrors.json
- include it in setup the same way as locale but in `share/pacman-mirrors`

philmmanjaro commented:
@fhdk: looks good. Package is now at 50b356b. You're simply to fast in developing.

fhdk commented:
:smile_cat: - fast? No - retired and sometimes bored out of my skull. Besides I like the challenge and the fact that I can put newly adapted knowledge into something useful.

fhdk commented:
I think it would be good if someone would sync the translation file with transifex

philmmanjaro commented:
I've updated the file on transifex. Seems some lines need to be translated ... Package is now at 4e2ac77

fhdk commented:
@philmmanjaro 
I think it is safe now, to let other testers have a try on a rebuild of this package.

Is it safe to remove the obsolete folders from the data folder?

fhdk commented:
@Huluti @Kirek @philmmanjaro 
How is it with the review?

philmmanjaro commented:
The package is now at 95ab000. However I had no time yet to do any code review so far. @Kirek and @Huluti. What do you guys say?

fhdk commented:
Been looking at the structure
`conf` installs to `/etc/pacman-mirrors.conf`
`locale` installs to `/usr/share/locale/*`
`scripts` installs to `/usr/bin/pacman-mirrors`
added
`pacman-mirrors` installs to `/usr/share/pacman-mirrors/mirrors.json`
removed `data`
```
├── conf
│   └── pacman-mirrors.conf
├── pacman-mirrors
│   └── mirrors.json
├── locale
├── scripts
│   └── pacman-mirrors
```
Is this acceptable?

fhdk commented:
@Huluti @Kirek @philmmanjaro 

And now all tests are green

Huluti commented:
I'm sorry, I've had very little free time lately. For me, the code looks good. Have you tested almost all available otpions? If so, I think we're good!

fhdk commented:
I have been through the code in - almost every possible way - and I am out of options to make it break. Need someone to do what I have not thought of :smile_cat:

fhdk commented:
I am humbly accepting that you have merged my branch.

Thank you for your trust in my small abilities to contribute to the Manjaro project.
